weasyl/controllers/user.py

```diff
@@ -65,6 +95,101 @@ def signin_post_(request):
+    if session_life > 300:
+        _cleanup_2fa_session()
+        return Response(define.errorpage(
+            request.userid,
+            errorcode.error_messages['TwoFactorAuthenticationAuthenticationTimeout'],
+            [["Sign In", "/signin"], ["Return to the Home Page", "/"]]))
+    else:
+        ref = request.params["referer"] if "referer" in request.params else "/"
+        return Response(define.webpage(
+            request.userid,
+            "etc/signin_2fa_auth.html",
+            [define.get_display_name(tfa_userid), ref, two_factor_auth.get_number_of_recovery_codes(tfa_userid),
+             None]))
+
+
+@guest_required
+@token_checked
+def signin_2fa_auth_post_(request):
+    sess = define.get_weasyl_session()
+
+    # Only render page if the password has been authenticated (we have a UserID stored in the session)
+    if '2fa_pwd_auth_userid' not in sess.additional_data:
+        return Response(define.errorpage(request.userid, errorcode.permission))
+    tfa_userid = sess.additional_data['2fa_pwd_auth_userid']
+
+    session_life = arrow.now().timestamp - sess.additional_data['2fa_pwd_auth_timestamp']
+    if session_life > 300:
+        # Maximum secondary authentication time: 5 minutes
+        _cleanup_2fa_session()
+        return Response(define.errorpage(
+            request.userid,
+            errorcode.error_messages['TwoFactorAuthenticationAuthenticationTimeout'],
+            [["Sign In", "/signin"], ["Return to the Home Page", "/"]]
+        ))
+    elif two_factor_auth.verify(tfa_userid, request.params["tfaresponse"]):
+        # 2FA passed, so login and cleanup.
+        _cleanup_2fa_session()
+        login.signin(tfa_userid)
+        ref = request.params["referer"] or "/"
+        # User is out of recovery codes, so force-deactivate 2FA
+        if two_factor_auth.get_number_of_recovery_codes(tfa_userid) == 0:
+            two_factor_auth.force_deactivate(tfa_userid)
+            return Response(define.errorpage(
+                tfa_userid,
+                errorcode.error_messages['TwoFactorAuthenticationZeroRecoveryCodesRemaining'],
+                [["2FA Dashboard", "/control/2fa/status"], ["Return to the Home Page", "/"]]
+            ))
+        # Return to the target page, restricting to the path portion of 'ref' per urlparse.
+        raise HTTPSeeOther(location=urlparse.urlparse(ref).path)
+    elif sess.additional_data['2fa_pwd_auth_attempts'] >= 5:
+        # Hinder brute-forcing the 2FA token or recovery code by enforcing an upper-bound on 2FA auth attempts.
+        _cleanup_2fa_session()
+        return Response(define.errorpage(
+            request.userid,
+            errorcode.error_messages['TwoFactorAuthenticationAuthenticationAttemptsExceeded'],
+            [["Sign In", "/signin"], ["Return to the Home Page", "/"]]
+        ))
+    else:
+        # Log the failed authentication attempt to the session and save
+        sess.additional_data['2fa_pwd_auth_attempts'] += 1
```

Comment:
@charmander How much do you know about our threading model with Pyramid and tornado? I'm vague on it.

This may be crazy, but what's to stop a user from issuing a hundred requests at once? This looks like a read-modify-write situation. There's not an easy way to mutex it because sessions are created on the way in and written back to the database before the response is returned to the user.